src/core/net/dns_client.cpp

```diff
@@ -815,20 +862,77 @@ void Client::SendQuery(Query &aQuery, QueryInfo &aInfo, bool aUpdateTimer)
         SuccessOrExit(error = message->Append(Question(kQuestionRecordTypes[aInfo.mQueryType][num])));
     }
 
-    messageInfo.SetPeerAddr(aInfo.mConfig.GetServerSockAddr().GetAddress());
-    messageInfo.SetPeerPort(aInfo.mConfig.GetServerSockAddr().GetPort());
+    length = message->GetLength() - message->GetOffset();
+#if OPENTHREAD_CONFIG_DNS_CLIENT_OVER_TCP_ENABLE
 
-    SuccessOrExit(error = mSocket.SendTo(*message, messageInfo));
+    // Send udp query in case of message being greater than kUdpQueryMaxSize or as last resort.
+    if (aInfo.mConfig.GetTransportProto() == QueryConfig::kDnsTcpTransport)
+    {
+        mSendLink.mNext = nullptr;
+        mSendLink.mData = mSendBufferBytes;
 
-exit:
-    FreeMessageOnError(message, error);
+        // Prepending the DNS query with length of the packet according to RFC1035.
+        WriteUint16(length, mSendBufferBytes);
+        message->Read(message->GetOffset(), (mSendBufferBytes + sizeof(uint16_t)), length);
+        mSendLink.mLength = length + sizeof(uint16_t);
 
-    UpdateQuery(aQuery, aInfo);
+        mTcpTimer.Stop();
+        switch (mTcpState)
+        {
+        case kTcpUninitialized:
+            mCurrentTcpQuery = &aQuery;
+            SuccessOrExit(InitTcpSocket());
+            SuccessOrExit(mEndpoint.Connect(AsCoreType(&aInfo.mConfig.mServerSockAddr), OT_TCP_CONNECT_NO_FAST_OPEN));
+            break;
+        case kTcpInitialized:
+            if (memcmp(&mEndpoint.GetPeerAddress(), &aInfo.mConfig.mServerSockAddr, sizeof(otSockAddr)) != 0)
+            {
+                error = kErrorFailed;
+                ExitNow();
+            }
+            mCurrentTcpQuery = &aQuery;
+            mTcpState        = kTcpSendingInProgress;
+            SuccessOrExit(error = mEndpoint.SendByReference(mSendLink, /* aFlags */ 0));
+            break;
+        case kTcpConecting:
+        case kTcpSendingInProgress:
+            if (memcmp(&mEndpoint.GetPeerAddress(), &aInfo.mConfig.mServerSockAddr, sizeof(otSockAddr)) != 0)
+            {
+                error = kErrorFailed;
+                ExitNow();
+            }
```

Comment:
In this case, we are waiting for the previous TCP tx to be done to send the next.
We can consider instead to add the new query to end of the `mSendBuffer`.
TCP provide mechanism to extend the last send buffer using `SendByExtension`. 

We can consider the following:
- If In `kTcpConecting`
   - We update `mSendLinkBuffer`, append new query at the end of previous one (write after `mSendLink.mLength` and update `mSendLink.mLength`)
   - The `mSendLink` will be passed to TCP once we get connection-established callback.
- If in `kTcpSending` (indicates `mSendLink` is in use by TCP)
   - We can append the new content at the end, after `mSendLink.mLength`
   - Then call `SendByExtension` passing in the number of appended bytes.
   - It is important to not change the `mSendLink.mLength` in this case (as `mSendLink` is owned by TCP and it will itself update the length in `SendByExtension`).
- If either case if buffer is full we return `kErrorNoBuf`